### PR TITLE
Hooks for C++ API

### PR DESCRIPTION
**Pull request by malvika2147**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24393 Hooks for C++ API**

Summary:
Ability to register hook on a variable, similar to python autograd API. register_hook will take a function as argument and create a CppFunctionPreHook similar to PyFunctionPreHook.
It will return the index of the hook which can be passed to remove_hook to disable the hook.

Test Plan:
Added tests.

Differential Revision: [D16861722](https://our.internmc.facebook.com/intern/diff/D16861722)

### CODE REVIEW

**Comment by ezyang** [comment]
Tests? :)

File: torch/csrc/autograd/variable.h [hunk 1]
```diff
@@ -291,6 +292,22 @@ struct TORCH_API Variable : public at::Tensor {
...
+  // Remove hook at given position
+  void remove_hook(unsigned pos);
+
+  // Returns the index of the hook in the list which can be used to remove hook
+  // Register a hook with no return value
+  template <typename T>
+  hook_return_void_t<T> register_hook(T&& hook);
+  // Register a hook with variable return value
+  template <typename T>
+  hook_return_var_t<T> register_hook(T&& hook);
```
Comment (ezyang) [inline torch/csrc/autograd/variable.h]: If my hook returns an undefined tensor, what happens?

File: torch/csrc/autograd/variable.h [hunk 2]
```diff
@@ -291,6 +292,22 @@ struct TORCH_API Variable : public at::Tensor {
...
+
+  void create_cpp_hook();
```
Comment (ezyang) [inline torch/csrc/autograd/variable.h]: This method probably shouldn't be public?

File: torch/csrc/autograd/cpp_hook.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#include <torch/csrc/autograd/cpp_hook.h>
+#include <torch/csrc/autograd/variable.h>
+#include <torch/csrc/autograd/custom_function.h>
+
+using torch::autograd::Variable;
+static void check_single_result (Variable value, Variable result, std::string hook_name);
```
Comment (ezyang) [inline torch/csrc/autograd/cpp_hook.cpp]: FYI it's more idiomatic in C++ to put static functions in an anonymous namespace

File: torch/csrc/autograd/cpp_hook.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+: hooks_(hooks)
+, value_idx_(value_idx)
+{}
+
+variable_list CppFunctionPreHook::operator()(const variable_list& values) {
+  auto value = values[value_idx_];
+
+  for (unsigned i = 0; i < hooks_->size(); ++i) {
+    auto &hook = (*hooks_)[i];
+    if (hook == NULL) {
```
Comment (ezyang) [inline torch/csrc/autograd/cpp_hook.cpp]: nit: nullptr

File: torch/csrc/autograd/variable.h (hunk 1 above)
Comment (ezyang) [reply torch/csrc/autograd/variable.h]: It looks like it errors. But maybe you should treat that as "no change requested"? (In analogy to returning `None` from a normal hook.)

File: torch/csrc/autograd/custom_function.cpp [hunk 5]
```diff
@@ -100,6 +100,33 @@ variable_list _wrap_outputs(const variable_list &input_vars,
...
+void check_variable_result(const Variable& original, const Variable& result, std::string hook_name) {
+  if (original.type() != result.type()) {
+    std::stringstream ss;
+    ss << "hook '" << hook_name << "' has changed the type of value (";
+    ss << "was " << original.toString() << " got ";
+    ss << result.toString() << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  if (original.is_cuda() != result.is_cuda()) {
```
Comment (ezyang) [inline torch/csrc/autograd/custom_function.cpp]: Hmm... where did you get this code from :o) Let's compare `device()` instead; that'll be more robust to non-CUDA/CPU things.

Comment (ezyang) [reply torch/csrc/autograd/custom_function.cpp]: Check that the `dtype()` is the same, that the `device()` is the same and the `layout()` is the same.

File: torch/csrc/autograd/cpp_hook.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace torch { namespace autograd {
+
+using hooks_list = std::vector<std::function<Variable(const Variable&)>>;
+
+struct CppFunctionPreHook : public FunctionPreHook {
+  CppFunctionPreHook(const std::shared_ptr<hooks_list> &hooks, int value_idx);
+  variable_list operator()(const variable_list& values) override;
+
+  std::shared_ptr<hooks_list> hooks_;
```
Comment (ezyang) [inline torch/csrc/autograd/cpp_hook.h]: Is there any reason this has to be a shared pointer to a vector, as opposed to just a vector?

File: torch/csrc/autograd/variable.cpp [hunk 7]
```diff
@@ -180,4 +180,21 @@ void Variable::rebase_history(Edge gradient_edge) {
...
+  std::unique_ptr<FunctionPreHook> hook_ptr(new CppFunctionPreHook(list, output_nr()));
+  auto fn = grad_fn();
+  if (fn) {
+    fn->add_pre_hook(std::move(hook_ptr));
+  }
+}
+
+void Variable::remove_hook(unsigned pos) {
+  auto &list = get_autograd_meta()->cpp_hooks_list;
+  TORCH_CHECK(list && pos < list->size() , "Invalid idx");
```
Comment (ezyang) [inline torch/csrc/autograd/variable.cpp]: A more descriptive error message will be appreciated by users all around the world :)

**Review by ezyang — DISMISSED** [review]
Looks legit but we shouldn't land this without tests

File: torch/csrc/autograd/custom_function.cpp (hunk 5 above)
Comment (malvika2147) [reply torch/csrc/autograd/custom_function.cpp]: This was just moved from python_hook.cpp

Comment (ezyang) [reply torch/csrc/autograd/custom_function.cpp]: OK. We can be semantics preserving for now, but that code should get fixed :) (File a bug, mayhap?)

**Comment by ezyang** [comment]
Needs a rebase past master failure

File: test/cpp/api/autograd.cpp [hunk 8]
```diff
@@ -432,6 +432,79 @@ TEST(CustomAutogradTest, ReentrantPriority) {
...
+  y.set_requires_grad(true);
+
+  int counter = 0;
+
+  std::function<void(int, Variable)> bw_hook([&counter](int inc, Variable grad){
+    counter += inc;
+  });
+
+  Variable z = x * x + x * 2 + x * y + y;
+  x.register_hook(std::bind(bw_hook, 0 ,std::placeholders::_1));
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: Unless necessary (e.g., moving arguments into a lambda), it's generally better not to use `std::bind`, and just pass in a lambda that does the binding.

**Review by ezyang — APPROVED** [review]

File: torch/csrc/autograd/cpp_hook.h (hunk 6 above)
Comment (malvika2147) [reply torch/csrc/autograd/cpp_hook.h]: The variable and all the CppFunctionHooks need to be able to access it. This is similar to what we do in python with the backward_hooks dict.

**Pull request closed without merging**